server/client/index.ts

```diff
@@ -0,0 +1,111 @@
+import { get } from "lodash";
+import {
+  IContextProvider,
+  ILegacyScopedClusterClient,
+  OpenSearchDashboardsRequest,
+  RequestHandler,
+  RequestHandlerContext,
+} from "opensearch-dashboards/server";
+import { IRequestHandlerContentWithDataSource, IGetClientProps, DashboardRequestEnhancedWithContext } from "./interface";
+
+export const getClientSupportMDS = (props: IGetClientProps) => {
+  const originalAsScoped = props.client.asScoped;
+  const handler: IContextProvider<RequestHandler<unknown, unknown, unknown>, "core"> = (
+    context: RequestHandlerContext,
+    request: OpenSearchDashboardsRequest
+  ) => {
+    (request as DashboardRequestEnhancedWithContext)[`${props.pluginId}_context`] = context as IRequestHandlerContentWithDataSource;
+    return {} as any;
+  };
+
+  /**
+   * asScoped can not get the request context,
+   * add _context to request
+   */
+  props.core.http.registerRouteHandlerContext(`${props.pluginId}_MDS_CTX_SUPPORT` as "core", handler);
+
+  /**
+   * it is not a good practice to rewrite the method like this
+   * but JS does not provide a method to copy a class instance
+   */
+  props.client.asScoped = function (request: DashboardRequestEnhancedWithContext): ILegacyScopedClusterClient {
```

Comment:
What do you think of using `Proxy` instead of overwrite the original method? Something like this:
```javascript
    const proxy = new Proxy(legacyClient, {
      get(target, prop) {
        if (prop === 'asScoped') {
          return (request) => {
            if (/* data source id presents */) {
              return {
                callAsCurrentUser: async () => {/.../},
                callAsInternalUser: async () => {/.../},
              };
            }
            return target.asScoped(request);
          };
        }
        return Reflect.get(...arguments);
      },
    });
    // then use legacyClient proxy
    const indexService = new IndexService(proxy);
```